lxd/storage/drivers/driver_pure_util.go

```diff
@@ -12,3 +49,186 @@ func newPureClient(driver *pure) *pureClient {
+		respBody = &pureError{}
+	}
+
+	// Extract the response body if requested.
+	if respBody != nil {
+		err = json.NewDecoder(resp.Body).Decode(respBody)
+		if err != nil {
+			return fmt.Errorf("Failed to read response body from %q: %w", path, err)
+		}
+	}
+
+	// Extract the response headers if requested.
+	if respHeaders != nil {
+		for k, v := range resp.Header {
+			respHeaders[k] = strings.Join(v, ",")
+		}
+	}
+
+	// Return the formatted error from the body
+	pureErr, ok := respBody.(*pureError)
+	if ok {
+		pureErr.StatusCode = resp.StatusCode
+		return pureErr
+	}
+
+	return nil
+}
+
+// requestAuthenticated issues an authenticated HTTP request against the PureStorage gateway. In case
+// the access token is expired, the function will try to obtain a new one.
+func (p *pureClient) requestAuthenticated(method string, path string, reqBody io.Reader, respBody any) error {
+	retries := 1
+	for {
+		// Ensure we are logged into the PureStorage.
+		err := p.login()
+		if err != nil {
+			return err
+		}
+
+		// Set access token as request header.
+		reqHeaders := map[string]string{
+			"X-Auth-Token": p.accessToken,
+		}
+
+		// Initiate request.
+		err = p.request(method, path, reqBody, reqHeaders, respBody, nil)
+		if err != nil {
+			if api.StatusErrorCheck(err, http.StatusUnauthorized) && retries > 0 {
+				// Access token seems to be expired.
+				// Reset the token and try one more time.
+				p.accessToken = ""
+				retries--
+				continue
+			}
+
+			// Either the error is not of type unauthorized or the maximum number of
+			// retries has been exceeded.
+			return err
+		}
+
+		return nil
+	}
+}
+
+// getAPIVersion returns the list of API version that are supported by the PureStorage.
+func (p *pureClient) getAPIVersions() ([]string, error) {
+	var resp struct {
+		APIVersions []string `json:"version"`
+	}
+
+	err := p.request(http.MethodGet, "/api/api_version", nil, nil, &resp, nil)
+	if err != nil {
+		return nil, fmt.Errorf("Failed retrieve available API versions from PureStorage: %w", err)
+	}
+
+	if len(resp.APIVersions) == 0 {
+		return nil, fmt.Errorf("PureStorage does not support any API version")
+	}
+
+	return resp.APIVersions, nil
+}
+
+// login initiates an authentication request against the PureStorage using the API token. If successful,
+// an access token is retrieved and stored within a client. The access token is then used for further
+// authentication.
+func (p *pureClient) login() error {
+	if p.accessToken != "" {
+		// Token has been already obtained.
+		return nil
+	}
+
+	reqHeaders := map[string]string{
+		"Api-Token": p.driver.config["pure.api.token"],
+	}
+
+	respHeaders := make(map[string]string)
+
+	err := p.request(http.MethodPost, "/login", nil, reqHeaders, nil, respHeaders)
+	if err != nil {
+		return fmt.Errorf("Failed to login: %w", err)
+	}
+
+	p.accessToken = respHeaders["X-Auth-Token"]
+	if p.accessToken == "" {
+		return errors.New("Failed to obtain access token")
```

Comment:
do we get any other sort of error message from this post?

Reply:
If we get any error from this post, it will be returned as "Failed to login: <err>".
For example: `Error: Failed to create storage pool "pure": Failed to login: Unable to list user for API token` where `Unable to list user for API token` is PureStorage error.

The check `p.accessToken == ""` is just to ensure that we also successfully parsed the the access token from the response. This code is only reached if Post request was successful.